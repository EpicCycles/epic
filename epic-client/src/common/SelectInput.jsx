import React, {Component} from "react";

/**
 * DIsplay generic select boc
 * @param props
 *  className - string optional
 *  error - string optional
 *  title - string optional
 *  label - string optional
 *  fieldName - String required
 *  onChange - function from container component
 *  isMultiple - boolean optional
 *  multipleSize - optional integer number shown - not used if isMultiple is not true
 *  value - Array of Strings for selected values optional
 *  isEmptyAllowed - boolean optional
 *  options - array:
 *      value - String - value for database
 *      name - String - value to be selected, optional value will be used if none passed
 *      isDefault - boolean optional
 *
 */
class SelectInput extends Component {
    buildOptions = (options, isEmptyAllowed, value) => {
        let displayOptions = [];
        let selectedValues = value? value : [];
        isEmptyAllowed && displayOptions.push({ value: "", name: "None", selected: (!value) });
        options.forEach((option) => {
            const displayName = option.name ? option.name : option.value;
            const displaySelected = selectedValues.length > 0 ? selectedValues.includes(option.value) : option.isDefault;
            displayOptions.push({ value: option.value, name: displayName, selected: displaySelected });
        });
        return displayOptions;
    };

    render() {
        const { className, fieldName, error, title, label, onChange, isMultiple, multipleSize, value, isEmptyAllowed, options } = this.props;
        const displayOptions = this.buildOptions(options, isEmptyAllowed, value);
         return <div
            id={"container" + fieldName}
            key={"container" + fieldName}
            className={className + (error ? " error" : "")}
            title={error}
        >
            {label && <label>{label}</label>}

            <select
                name={fieldName}
                id={fieldName}
                key={fieldName}
                title={title}
                multiple={isMultiple}
                size={multipleSize}
                onChange={event => onChange(event.target.name, event.target.value)}
            >
                {displayOptions.map((option) => {
                    return <option
                        id={`${fieldName}_${option.value}`}
                        key={`${fieldName}_${option.value}`}
                        value={option.value}
                        selected={option.selected}
                    >
                        {option.name}
                    </option>
                })}
            </select>
        </div>;
    };
}
export default SelectInput;