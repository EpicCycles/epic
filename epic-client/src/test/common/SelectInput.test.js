import React from 'react';
import SelectInput from "../../common/SelectInput";

test("it renders with minimum data", () => {
    const options = [{ value: "H", name: "Home" }];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
    />);
    expect(component).toMatchSnapshot();
});

test("it renders an empty option if one is allowed", () => {
    const options = [{ value: "H", name: "Home" }];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={true}
    />);
    expect(component).toMatchSnapshot();
    expect(component.find('option').length).toBe(2);
});
test("it renders a multiple select array if one is passed", () => {
    const options = [
        { value: "H", name: "Home" },
        { value: "A", name: "Away" },
        { value: "O", name: "Other" },
    ];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={true}
        isMultiple={true}
    />);
    expect(component).toMatchSnapshot();
    expect(component.find('option').length).toBe(4);
});
test("it renders a multiple select array if one is passed with correct size", () => {
    const options = [
        { value: "H", name: "Home" },
        { value: "A", name: "Away" },
        { value: "O", name: "Other" },
    ];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={true}
        multipleSize={2}
    />);
    expect(component).toMatchSnapshot();
    expect(component.find('option').length).toBe(3);
});
test("it sets a value to selected if it is in the array of selected values - single select", () => {
    const options = [
        { value: "H", name: "Home" },
        { value: "A", name: "Away" },
        { value: "O", name: "Other" },
    ];
    const selectedValue = ["A"];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={true}
        isMultiple={true}
        value={selectedValue}
    />);
    expect(component).toMatchSnapshot();
});
test("it sets a value to selected if it is in the array of selected values - multiple select", () => {
    const options = [
        { value: 1, name: "Home" },
        { value: 2, name: "Away" },
        { value: 3, name: "Other" },
    ];
    const selectedValue = [1,3];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={true}
        value={selectedValue}
    />);
    expect(component).toMatchSnapshot();
});
test("it renders the default value as selected if no value is passed", () => {
    const options = [
        { value: "H", name: "Home" },
        { value: "A", name: "Away", isDefault: true },
        { value: "O", name: "Other" },
    ];
    const selectedValue = [];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={false}
        value={selectedValue}
    />);
    expect(component).toMatchSnapshot();
});
test("it renders an error, title and a passed class", () => {
    const options = [
        { value: "H", name: "Home" },
        { value: "A", name: "Away", isDefault: true },
        { value: "O", name: "Other" },
    ];
    const selectedValue = [];
    const component = shallow(<SelectInput
        fieldName="myField"
        onChange={jest.fn()}
        options={options}
        isEmptyAllowed={false}
        isMultiple={false}
        value={selectedValue}
        error="show an error"
        title="show a title"
        className="pink"
    />);
    expect(component).toMatchSnapshot();
});
