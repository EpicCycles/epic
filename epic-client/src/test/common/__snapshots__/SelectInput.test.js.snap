// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a multiple select array if one is passed 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_"
      key="myField_"
      selected={true}
      value=""
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`it renders a multiple select array if one is passed with correct size 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
    size={2}
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`it renders an empty option if one is allowed 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_"
      key="myField_"
      selected={true}
      value=""
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
  </select>
</div>
`;

exports[`it renders an error, title and a passed class 1`] = `
<div
  className="pink error"
  id="containermyField"
  key="containermyField"
  title="show an error"
>
  <select
    id="myField"
    key="myField"
    multiple={false}
    name="myField"
    onChange={[Function]}
    title="show a title"
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      selected={true}
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`it renders the default value as selected if no value is passed 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    multiple={false}
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      selected={true}
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`it renders with minimum data 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_H"
      key="myField_H"
      value="H"
    >
      Home
    </option>
  </select>
</div>
`;

exports[`it sets a value to selected if it is in the array of selected values - multiple select 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_H"
      key="myField_H"
      selected={false}
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      selected={true}
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      selected={true}
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;

exports[`it sets a value to selected if it is in the array of selected values - single select 1`] = `
<div
  className="undefined"
  id="containermyField"
  key="containermyField"
>
  <select
    id="myField"
    key="myField"
    multiple={true}
    name="myField"
    onChange={[Function]}
  >
    <option
      id="myField_"
      key="myField_"
      selected={false}
      value=""
    >
      None
    </option>
    <option
      id="myField_H"
      key="myField_H"
      selected={false}
      value="H"
    >
      Home
    </option>
    <option
      id="myField_A"
      key="myField_A"
      selected={true}
      value="A"
    >
      Away
    </option>
    <option
      id="myField_O"
      key="myField_O"
      selected={false}
      value="O"
    >
      Other
    </option>
  </select>
</div>
`;
